certgen/certgen.go

```diff
@@ -0,0 +1,171 @@
+
+	// Optional fields
+
+	// Prefix is a prefix that will be used in conjunction with the
+	// SubjectCommonName to create a more specific DNS SAN.
+	Prefix string
+	// NotAfterInterval specifies an interval in which the NotAfter time of a
+	// certificate must fall.
+	NotAfterInterval *interval.Interval
+}
+
+// CA is a Certificate Authority that issues synthetic certificates and
+// certificate chains using its RootCert and RootKey.
+type CA struct {
+	RootCert   *x509.Certificate
+	RootKey    crypto.Signer
+	CertConfig CertificateConfig
+}
+
+// IssueCertificate creates a new leaf certificate, issued by the root specified
+// in the RootCert and RootKey fields of the CA, and configured using the
+// CertConfig in the CA.
+func (ca *CA) IssueCertificate() (*x509.Certificate, error) {
+	// TODO(katjoyce): Make the type of key generated configurable.
+	key, err := rsa.GenerateKey(crand.Reader, keySize)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key pair: %s", err)
+	}
+
+	template, err := leafTemplate(ca.CertConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf template: %s", err)
+	}
+
+	leafDER, err := x509.CreateCertificate(crand.Reader, template, ca.RootCert, key.Public(), ca.RootKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf certificate: %s", err)
+	}
+
+	leaf, err := x509.ParseCertificate(leafDER)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing leaf certificate DER: %s", err)
+	}
+
+	return leaf, nil
+}
+
+func leafTemplate(c CertificateConfig) (*x509.Certificate, error) {
+	sn, err := randSerialNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	notAfter := randNotAfter(c.NotAfterInterval)
+
+	return &x509.Certificate{
+		SerialNumber: sn,
+		Subject: pkix.Name{
+			Country:            []string{c.SubjectCountry},
+			Organization:       []string{c.SubjectOrganization},
+			OrganizationalUnit: []string{c.SubjectOrganizationalUnit},
+			Locality:           []string{c.SubjectLocality},
+			CommonName:         c.SubjectCommonName,
+		},
+		NotBefore:          notAfter.Add(-certValidity),
+		NotAfter:           notAfter,
+		SignatureAlgorithm: c.SignatureAlgorithm,
+
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+		DNSNames:              []string{c.SubjectCommonName, extendedDNSSAN(c.Prefix, c.SubjectCommonName)},
+	}, nil
+}
+
+func randSerialNumber() (*big.Int, error) {
+	i := big.NewInt(0)
+	return crand.Int(crand.Reader, i.SetUint64(math.MaxUint64))
+}
+
+func randNotAfter(notAfterInterval *interval.Interval) time.Time {
+	if notAfterInterval == nil {
+		// TODO(katjoyce): Make this random.
+		return timeNowUTC().Add(certValidity)
+	}
+
+	rand.Seed(timeNowUTC().UnixNano())
+
+	start := notAfterInterval.Start.Unix()
+	end := notAfterInterval.End.Unix()
+	delta := end - start
```

Comment:
Will delta generally differ from certValidity?  In the leafTemplate function I see that you subtract certValidity from the time returned by randNotAfter to obtain a not-before timestamp.

It's not very clear to me what benefit the interval.Interval brings here. (sorry)

Reply:
Hehe, don't be sorry!  If it's not clear, I need to change it to be clearer!

As mentioned above, having discussed in person I have added a comment.  See what you think :)